### ISSUE TITLE
Feature request: Export metrics to Prometheus

### ISSUE THREAD

**Issue by qdm12** [passteque/gluetun#279]
1. What's the feature?

- Export metrics to Prometheus on:
    - Scrap iptables data to get network information (blocked data, vpn'ed data etc.)
    - Health information, such as when connectivity drops
    - VPN server information
    - the Go written Shadowsocks
    - the (Go) HTTP proxy
    - the Go written DNS server (blocked by #137)
    - ~the numbers found from #251~
    - Maybe add some bandwidth tests but that depends on the test server, so maybe not
- ~Add instructions on how to set this up with Grafana to have nice graphs~ See https://github.com/qdm12/gluetun-wiki/blob/latest/setup/advanced/metrics.md
- ~Add a Grafana dashboard example to the repository~ See https://github.com/passteque/gluetun/blob/master/metrics/grafana.json

2. Why do you need this feature?

Because metrics are fun

3. Extra information?

**Comment by dorsegal** [comment]
That would be awesome!

**Comment by SR-G** [comment]
So any news on this ?
I would also find this very helpful. Especially the "connection lost", "connected country", ... informations.

Maybe "start small" / do "baby steps" regarding this new evolution, as the feature list is quite long and may be overhelming ?

What would be the "architecture" for these metrics ? Something embedded inside gluetun code itself (and exposed by ... the control server, on a /metrics path, maybe ?), or an external separated program (a `gluetun-exporter` go standalone project, able to externally gather (in some way) all the metrics) ?

**Comment by TheGameProfi** [comment]
Hey everyone 👋

Since there hasn’t been much activity or communication lately, I went ahead and created a Prometheus Exporter myself.
It’s still a work in progress — I’m currently experimenting with getting it to auto-configure, but the core functionality is already there.

Feel free to check it out and let me know what you think:
👉 [Gluetun-Exporter](https://github.com/thecfu/gluetun-exporter)

I’m totally open to feedback, ideas, or suggestions! 😊

**Comment by SR-G** [comment]
Interesting, thanks !

But at very quick glance i'm not so sure how it has to be launched / what is the principe (the README.md is not 100% clear about that), especially with Standalone mode.

I build the container i executed it with : 

```
docker run --rm --name "gluetun-exporter" \
  -e"GLUETUN_HOST=localhost" \ <-- same with "gluetun"
  -e"GLUETUN_PORT=8000" \
  -e"EXPORTER_PORT=8001" \
  -e"EXPORTER_ROLE=gluetun" \
  -e"EXPORTER_INTERVAL=30" \
  -e"GLUETUN_USERNAME=gluetun" \
  -e"GLUETUN_PASSWORD=<password>" \
  registry.tensin.org/gluetun-exporter
2025-04-08T20:24:36Z INFO [gluetun-exporter] Starting prometheus exporter...
2025-04-08T20:24:36Z INFO [gluetun-exporter] Registering metrics...
2025-04-08T20:25:36Z INFO [gluetun-exporter] Updating Metrics...
```

And i have nothing exposed on port :8001 ... (and nothing displayed in logs)

So : 
- should the "gluetun-exported" be linked against the gluetun container or not (with `--network=container:gluetun`, and then by exposing the `gluetun-exporter` port from the gluetun container (like for any other services) ?
- what is the hostname / port about, is it the GLUETUN API ? (i expose it through a load balancer with an HTTPS URL, but then i'm confused by the "host" entry and not "URL with protocol")

Thanks ;)

**Comment by TheGameProfi** [comment]
Hey there, 
From the first look you didn't exposed the exporter ip via `-p 8001:8001` or you add the network part and exposes it via gluetun.
I just set it up and its not really finished, so currently its only via http connection to the host and port of the gluetun api endpoint.
Bu ttomorrow I will try to make it possible via the full url and update the Readme.
Also I think it would be better to not spam this Issue and instead open another Issue on the exporter Repo, for more questions.

**Comment by SR-G** [comment]
Hey. 
What a stupid mistake, i was probably too quick on this and confused by the "network link or not" question (if network link, the ports of the extra containers are not exposed directly, but through the gluetun container).
Indeed with the port, the /metrics is exposed (so it has definitely to be executed without any network link on the gluetun container, got it).
Thanks.

**Comment by jcheroske** [comment]
Is a metrics endpoint on the roadmap for Gluetun?

**Comment by qdm12** [comment]
It will come, but probably ~March 2026, there are some prerequirments I want done before. The new dns system has already built-in support for metrics/prometheus, but it's not used yet, so some work has been done even if it cannot be seen yet 😉

**Comment by willianpaixao** [comment]
@qdm12 if you define a fine grained set of issues and their DoD, I can  tackle a few of them.

**Comment by qdm12** [comment]
Some progress done with some initial metrics (vpn link bandwidth and total date sent/received), see https://github.com/qdm12/gluetun-wiki/blob/latest/setup/advanced/metrics.md

It works fine with Prometheus + Grafana. A bunch more metrics to come soon, after the v3.42.0 release.

**Comment by SR-G** [comment]
Promising ! 

For the next metrics to be added, i'd expect to be able to retrieve (and to display in a GRAFANA dashboard, etc.) that kind of metrics : 
- type of connectivity (VPW, WIREGUARD)
- status of the connectivity (up, down)
- current IP (as a string)
- time since start of the process
- time since last disconnect/reconnect of the VPN/WIREGUARD connectivity
- number of disconnects since start of the process
- number of connected clients/applications (in my case for example i have multiple docker containers = multiple processes routed through gluetun) (maybe per ports ?)
- bandwidth : instantaneous + cumulative (however cumulative can also be aggregated at grafana level, but maybe it's easier / more robust to expose it - since last restart - through gluetun), and up / down
- maybe for later : some metrics about the HTTP API
